phoenix-core/src/main/java/org/apache/phoenix/index/GlobalIndexChecker.java

```diff
@@ -0,0 +1,426 @@
+            Result result = region.get(get);
+            if (result.isEmpty()) {
+                return true;
+            }
+            if (Bytes.compareTo(result.getValue(emptyCF, emptyCQ), 0, FALSE_BYTES.length,
+                    FALSE_BYTES, 0, FALSE_BYTES.length) == 0) {
+                return false;
+            }
+            return true;
+        }
+
+        private boolean verifyRowAndRemoveVerifiedCell(List<Cell> cellList) throws IOException {
+            boolean valid = true;
+            boolean verifiedColumnExist = false;
+            long verifiedTs = 0;
+            long deleteTs = 0;
+            long maxTs;
+            long ts = 0;
+            long cellListSize = cellList.size();
+            Cell cell = null;
+            maxTs = 0;
+            if (cellListSize == 0) {
+                return true;
+            }
+            Iterator<Cell> cellIterator = cellList.iterator();
+            while (cellIterator.hasNext()) {
+                cell = cellIterator.next();
+                ts = cell.getTimestamp();
+                if (ts > maxTs) {
+                    maxTs = ts;
+                }
+                if (cell.getType() != Cell.Type.Put) {
+                    if (Bytes.compareTo(cell.getFamilyArray(), cell.getFamilyOffset(), cell.getFamilyLength(),
+                            emptyCF, 0, emptyCF.length) == 0) {
+                        if (ts > deleteTs) {
+                            deleteTs = ts;
+                        }
+                    }
+                    continue;
+                }
+                if (isVerifiedCell(cell)) {
+                    verifiedColumnExist = true;
+                    if (Bytes.compareTo(cell.getValueArray(), cell.getValueOffset(), cell.getValueLength(),
+                            FALSE_BYTES, 0, FALSE_BYTES.length) == 0) {
+                        if (verifiedTs < ts) {
+                            valid = false;
+                            verifiedTs = ts;
+                        }
+                    } else {
+                        if (verifiedTs < ts) {
+                            valid = true;
+                            verifiedTs = ts;
+                        }
+                    }
+                    if (cellListSize > 1) {
+                        cellIterator.remove();
+                    }
+                }
+            }
+            if (deleteTs >= maxTs) {
+                return true;
+            }
+            if (!verifiedColumnExist) {
+                byte[] rowKey = new byte[cell.getRowLength()];
+                System.arraycopy(cell.getRowArray(), cell.getRowOffset(), rowKey, 0, cell.getRowLength());
+                valid = verifyRow(rowKey, maxTs);
+            }
+            if (!valid || (verifiedColumnExist && verifiedTs != maxTs)) {
+                return false;
+            }
+            return true;
+        }
+
+        private long getMaxTimestamp(List<Cell> cellList) {
```

Comment:
Indexer.java has the same private function. Do you want to turn that one into public static and reuse here?

Reply:
These functions are not the same. One takes a Mutation and the other List<Cell>. Their implementations are also different.